core/src/main/java/com/linecorp/armeria/client/RedirectConfigBuilder.java

```diff
@@ -0,0 +1,159 @@
+import java.util.Set;
+import java.util.function.BiPredicate;
+
+import javax.annotation.Nullable;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
+
+import com.linecorp.armeria.common.annotation.UnstableApi;
+
+/**
+ * A builder for {@link RedirectConfig}.
+ */
+@UnstableApi
+public final class RedirectConfigBuilder {
+
+    @VisibleForTesting
+    static final BiPredicate<ClientRequestContext, String> allowAllDomains = (ctx, domain) -> true;
+
+    static final BiPredicate<ClientRequestContext, String> allowSameDomain = (ctx, domain) -> {
+        final Endpoint endpoint = ctx.endpoint();
+        if (endpoint == null) {
+            return false;
+        }
+        return endpoint.host().contains(domain);
+    };
+
+    private int maxRedirects = 19; // Widely used default value. https://stackoverflow.com/a/36041063/1736581
+
+    private boolean isAllowingAllDomains;
+    @Nullable
+    private Set<String> allowedDomains;
+    @Nullable
+    private BiPredicate<ClientRequestContext, String> predicate;
+
+    RedirectConfigBuilder() {}
+
+    /**
+     * Sets the maximum number of automatic redirection that the client executes.
+     */
+    public RedirectConfigBuilder maxRedirects(int maxRedirects) {
+        checkArgument(maxRedirects > 0, "maxRedirects: %s (expected: > 0)", maxRedirects);
+        this.maxRedirects = maxRedirects;
+        return this;
+    }
+
+    /**
+     * Allows automatic redirection for all domains.
+     * If the {@link WebClient} is created without a base URI, the {@link WebClient} executes
+     * automatic redirection for all domains by default. If the {@link WebClient} is created with a base URI,
+     * automatic redirection is executed for the domain of the base URI by default.
+     */
+    public RedirectConfigBuilder allowAllDomains() {
+        isAllowingAllDomains = true;
+        return this;
+    }
+
+    /**
+     * Sets the domains that are allowed for automatic redirection.
+     * If the {@link WebClient} is created without a base URI, the {@link WebClient} executes
+     * automatic redirection for all domains by default. If the {@link WebClient} is created with a base URI,
+     * automatic redirection is executed for the domain of the base URI by default.
+     */
+    public RedirectConfigBuilder allowDomains(String... domains) {
+        return allowDomains(ImmutableList.copyOf(requireNonNull(domains, "domains")));
+    }
+
+    /**
+     * Sets the domains that are allowed for automatic redirection.
+     * If the {@link WebClient} is created without a base URI, the {@link WebClient} executes
+     * automatic redirection for all domains by default. If the {@link WebClient} is created with a base URI,
+     * automatic redirection is executed for the domain of the base URI by default.
+     */
+    public RedirectConfigBuilder allowDomains(Iterable<String> domains) {
```

Comment:
Hm, is this a mutually exclusive condition? I thought it's rather `allowAllDomains` includes `allowDomains(...)` so that I can silently ignore it. Do you have any concerns if we allow it?